torchao/prototype/quantized_training/int8.py

```diff
@@ -0,0 +1,247 @@
+    dtype = kwargs.get("dtype", args[0].dtype)
+    device = kwargs.get("device", args[0].device)
+    return torch.zeros(args[0].shape, dtype=dtype, device=device)
+
+
+# out-of-place math ops always return plain tensor
+@Int8QTLinearWeight.implements([aten.sub.Tensor, aten.mul.Tensor])
+def _(func, types, args, kwargs):
+    args = [x.dequantize() if isinstance(x, Int8QTLinearWeight) else x for x in args]
+    return func(*args, **kwargs)
+
+
+@Int8QTLinearWeight.implements(aten.copy_.default)
+def _(func, types, args, kwargs):
+    if isinstance(args[0], Int8QTLinearWeight) and isinstance(args[1], Int8QTLinearWeight):
+        args[0].int_data.copy_(args[1].int_data, **kwargs)
+        args[0].scale.copy_(args[1].scale, **kwargs)
+
+    elif isinstance(args[0], Int8QTLinearWeight):
+        int_data, scale = Int8QTLinearWeight.quantize(args[1], stochastic_rounding=True)
+        args[0].int_data.copy_(int_data, **kwargs)
+        args[0].scale.copy_(scale, **kwargs)
+
+    else:
+        args[0].copy_(args[1].dequantize(), **kwargs)
+
+    return args[0]
+
+
+@Int8QTLinearWeight.implements([aten.addcdiv_.default, aten.add_.Tensor])
+def _(func, types, args, kwargs):
+    original = args[0]
+    out = func(args[0].dequantize(), *args[1:], **kwargs)
+    return original.copy_(out)
+
+
+# FSDP ops
+@Int8QTLinearWeight.implements(aten.split.Tensor)
+def _(func, types, args, kwargs):
+    if len(args) == 3 and args[2] != 0:
+        raise NotImplementedError("Int8QTLinearWeight only supports split at dim=0")
+
+    int8_weight: Int8QTLinearWeight = args[0]
+    int_data_list = func(int8_weight.int_data, *args[1:], **kwargs)
+    scale_list = func(int8_weight.scale, *args[1:], **kwargs)
+
+    out = [Int8QTLinearWeight(int_data, scale) for int_data, scale in zip(int_data_list, scale_list)]
+    return out
+
+
+@Int8QTLinearWeight.implements(aten.new_zeros.default)
+def _(func, types, args, kwargs):
+    size = args[1]
+    if len(size) != 2:
+        raise NotImplementedError
+
+    # TODO: handle pin_memory kwarg?
+    device = kwargs.get("device", args[0].device)
+    dtype = kwargs.get("dtype", args[0].dtype)
+    int_data = torch.zeros(size, device=device, dtype=torch.int8)
+    scale = torch.zeros(size[0], device=device, dtype=dtype)
+    return Int8QTLinearWeight(int_data, scale)
+
+
+# don't do anything. workaround for FSDP2. might give unexpected or wrong results.
+@Int8QTLinearWeight.implements([aten.view.default, aten.as_strided.default])
+def _(func, types, args, kwargs):
+    out = Int8QTLinearWeight(args[0].int_data, args[0].scale)
+    return return_and_correct_aliasing(func, args, kwargs, out)
+
+
+def int8_weight_only_quantized_training():
+    def apply_int8_linear_weight(linear: nn.Linear):
+        linear.weight = nn.Parameter(
+            Int8QTLinearWeight.from_float(linear.weight),
+            requires_grad=linear.weight.requires_grad,
+        )
+        return linear
+
+    return apply_int8_linear_weight
```

Comment:
I think you need `_get_linear_subclass_inserter` here

Reply:
`_get_linear_subclass_inserter()` will set `requires_grad=False`

https://github.com/pytorch/ao/blob/e7fc0ed0f2e130e44d250541aba71c7a7a985f15/torchao/quantization/quant_api.py#L261-L266

We can add `requires_grad` as extra argument to `_get_linear_subclass_inserter()`, but considering this is prototype/experimental, I would say leave it for future PR. Wdyt?

Reply:
Ok, add a TODO?

Reply:
Ok, add a TODO comment?